### PR TITLE
Add scenarios api

### PR DESCRIPTION
**Pull request by ktbee**
@gnarf @evmiguel here is my first pass at the scenarios API! I believe @gnarf has some thoughts on a SQL update helper, but let me know what you think so far. I've been testing it with postman. The endpoints will live at localhost:5000/api/scenarios. 

Let me know if you have any questions!

### CODE REVIEW

File: src/server/service/scenarios/db.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    return withClient(async client => {
+        const results = await client.query(
+            sql` SELECT * FROM scenario WHERE id = ${scenarioId};`
+        );
+        return results.rows[0];
+    });
+};
+
+exports.addScenario = async function addScenario(authorId, title, description) {
+    return withClientTransaction(async client => {
```
Comment (gnarf) [inline src/server/service/scenarios/db.js]: no need to transaction a single statement AFAIK

File: src/server/service/scenarios/endpoints.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+exports.addScenario = asyncMiddleware(exports.addScenarioAsync);
+
+exports.setScenarioAsync = async function(req, res) {
+    const { author_id, title, description } = req.body;
+    const scenarioId = req.params.scenario_id;
+
+    if (!author_id && !title && !description) {
+        const scenarioUpdateError = new Error(
+            "The scenario's title and description must be provided by a valid user"
```
Comment (gnarf) [inline src/server/service/scenarios/endpoints.js]: "Must provide author_id, title, or description to update"

File: src/server/service/scenarios/endpoints.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        res.send(result);
+    } catch (apiError) {
+        const error = new Error('Error while deleting scenario');
+        error.status = 500;
+        error.stack = apiError.stack;
+        throw error;
+    }
+};
+
+exports.deleteScenario = asyncMiddleware(exports.deleteScenarioAsync);
```
Comment (gnarf) [inline src/server/service/scenarios/endpoints.js]: this pattern is kind of getting to me at this point....

we should maybe just write these as `exports.deleteScenario = asyncMiddleware(async function deleteScenario(req, res, blah) {});`

File: src/server/service/scenarios/index.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    deleteScenario
+} = require('./endpoints.js');
+
+scenariosRouter.get('/:scenario_id', [getScenario]);
+
+scenariosRouter.put('/', [validateRequestBody, addScenario]);
+
+scenariosRouter.post('/:scenario_id', [validateRequestBody, setScenario]);
+
+scenariosRouter.delete('/:scenario_id', [validateRequestBody, deleteScenario]);
```
Comment (gnarf) [inline src/server/service/scenarios/index.js]: <3

File: src/server/util/db.js [hunk 5]
```diff
@@ -22,5 +22,5 @@ exports.usingTransaction = async (client, method) => {
     }
 };
 
-exports.withClientTransacation = method =>
+exports.withClientTransaction = method =>
```
Comment (gnarf) [inline src/server/util/db.js]: heh -- was it used anywhere else? 😳

File: src/server/service/scenarios/db.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
+const { sql } = require('../../util/sqlHelpers');
+const { withClient, withClientTransaction } = require('../../util/db');
+
+exports.getScenario = async function getScenario(scenarioId) {
+    return withClient(async client => {
+        const results = await client.query(
+            sql` SELECT * FROM scenario WHERE id = ${scenarioId};`
+        );
+        return results.rows[0];
+    });
```
Comment (gnarf) [inline src/server/service/scenarios/db.js]: this pattern repeats a lot too...

what about a thing to get rid of the `withClient` wrappings...

util/db.js:
```js
exports.query = (...args) => exports.withClient(client => client.query(...args));
```

lets you rewrite this:

```js
const results = await client.query(
  sql` SELECT * FROM scenario WHERE id = ${scenarioId};`
);
return results.rows[0];
```

if you want to get super terse -- 

```js
exports.getScenario = async scenarioId =>
    (await query(sql` SELECT * FROM scenario WHERE id = ${scenarioId};`))
        .rows[0];
```

File: src/server/util/db.js (hunk 5 above)
Comment (ktbee) [reply src/server/util/db.js]: Oh good call! It's in the roles db. I'll do a quick PR to fix that before merging this one in.

File: src/server/service/scenarios/db.js (hunk 1 above)
Comment (ktbee) [reply src/server/service/scenarios/db.js]: I think you're right! From this doc:

>PostgreSQL actually treats every SQL statement as being executed within a transaction. If you do not issue a BEGIN command, then each individual statement has an implicit BEGIN and (if successful) COMMIT wrapped around it. A group of statements surrounded by BEGIN and COMMIT is sometimes called a transaction block.

https://www.postgresql.org/docs/9.1/tutorial-transactions.html

File: src/server/service/scenarios/db.js (hunk 6 above)
Comment (ktbee) [reply src/server/service/scenarios/db.js]: Really cool!

File: src/server/service/scenarios/db.js [hunk 7]
```diff
@@ -1,53 +1,32 @@
...
-    author_id = ${newAuthorId},
-    title = ${newTitle},
-    description = ${newDescription}
-WHERE id = ${scenarioId}
-RETURNING *;
-        `);
-        return result.rows[0];
-    });
+    const result = await query(
+        updateQuery('scenario', `id = ${scenarioId}`, scenarioData)
```
Comment (gnarf) [inline src/server/service/scenarios/db.js]: this should just be `scenarioId` -- the updateQuery assumes if you pass in a number it's the `id`

File: src/server/service/scenarios/endpoints.js [hunk 8]
```diff
@@ -30,6 +28,7 @@ exports.addScenario = asyncMiddleware(async function addScenarioAsync(
 
     if (!userId || !title || !description) {
         const scenarioCreateError = new Error(
+            // eslint-disable-next-line quotes
```
Comment (gnarf) [inline src/server/service/scenarios/endpoints.js]: was this an error between eslint/prettier?  curious why it's double quoted in the first place.

Comment (ktbee) [reply src/server/service/scenarios/endpoints.js]: Yeah, they seemed to disagree and wanted me to use:

```
`"The scenario's title and description must be provided by a valid user"` 
```

Using single quotes tripped up prettier for some reason :/

**Pull request merged**